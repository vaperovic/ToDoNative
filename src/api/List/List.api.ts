import { AsyncStorage } from 'react-native';
import { List, ListInterface } from '../../models/List/List.model';

export interface ListsApiInterface {
    getLists: () => Promise<ListInterface[]>;
    saveLists: (lists: ListInterface[]) => Promise<ListInterface[]>;
}

interface ListObjectFromDb {
    name: string;
    index: number;
}

export class ListApi implements ListsApiInterface {
    private readonly url: string;

    constructor(url: string, path: string) {
        this.url = `${url}:${path}`;
    }

    public async getLists(): Promise<ListInterface[]> {
        try {
            const listsJson = await AsyncStorage.getItem(this.url);

            if (listsJson) {
                const parsedLists = JSON.parse(listsJson);

                return parsedLists.map((list: ListObjectFromDb) => {
                    return new List(
                        list.name,
                        list.index,
                    );
                });
            }

            return [];
        } catch (e) {
            console.log(e);
            return [];
        }
    }

    // This would normally be in 3 parts (POST, PUT and DELETE), but since I'm faking
    // api by using AsyncStorage and always saving whole list, I'll simplify it by
    // only using this function
    public async saveLists(lists: ListInterface[]): Promise<ListInterface[]> {
        try {
            const listsAsJson = JSON.stringify(lists.map((list: ListInterface) => {
                return list.toJSON();
            }));

            await AsyncStorage.setItem(this.url, listsAsJson);

            // ideally api would return new state or some sort of response to confirm
            // action, so I just return state of lists I just saved
            return lists;
        } catch (e) {
            console.log(e);
            return lists;
        }
    }

}
