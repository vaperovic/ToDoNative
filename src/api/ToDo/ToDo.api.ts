import { AsyncStorage } from "react-native";
import { ToDo, ToDoInterface } from "../../models/ToDo/ToDo.model";

export interface ToDoApiInterface {
    getToDos: () => Promise<ToDoInterface[]>;
    saveToDos: (todos: ToDoInterface[]) => Promise<ToDoInterface[]>;
}

interface ToDoDbObject {
    name: string;
    listId: number;
    completed: boolean;
    index: number;
}

export class ToDoApi<ToDoApiProps> implements ToDoApiInterface {
    private readonly url: string;

    constructor(url: string, path: string) {
        this.url = `${url}:${path}`;
    }

    public async getToDos(): Promise<ToDoInterface[]> {
        try {
            const todosJson = await AsyncStorage.getItem(this.url);

            if (todosJson) {
                const parsedTodos = JSON.parse(todosJson);

                return parsedTodos.map((todo: ToDoDbObject) => {
                    return new ToDo(
                        todo.name,
                        todo.listId,
                        todo.completed,
                        todo.index
                    )
                })
            }

            return [];
        } catch (e) {
            console.log(e);
            return [];
        }
    }

    // This would normally be in 3 parts (POST, PUT and DELETE), but since I'm faking
    // api by using AsyncStorage and always saving whole list, I'll simplify it by
    // only using this function
    public async saveToDos(todos: ToDoInterface[]): Promise<ToDoInterface[]> {
        try {
            const todosAsJson = JSON.stringify(todos.map((todo: ToDoInterface) => {
                return todo.toJSON();
            }));

            await AsyncStorage.setItem(this.url, todosAsJson);

            // ideally api would return new state or some sort of response to confirm
            // action, so I just return state of todos I just saved
            return todos;
        } catch (e) {
            console.log(e);
            return [];
        }
    }

}