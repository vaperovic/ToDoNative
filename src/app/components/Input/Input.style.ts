import { StyleSheet } from 'react-native';

export const InputStyle = StyleSheet.create({
  inputContainer: {
      display: 'flex',
      flexDirection: 'row',
      alignItems: 'center',
      borderTopWidth: 1,
      borderTopColor: '#eee',
    },
  addButton: {
      flex: 2,
      borderLeftColor: '#eee',
      borderLeftWidth: 1,
      display: 'flex',
      flexDirection: 'row',
      alignItems: 'center',
      paddingLeft: 10,
      height: '70%',
    },
  inputTextInput: {
      flex: 8,
    },
});
