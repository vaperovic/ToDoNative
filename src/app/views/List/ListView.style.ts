import { StyleSheet } from 'react-native';

export const ListViewStyle = StyleSheet.create({});
