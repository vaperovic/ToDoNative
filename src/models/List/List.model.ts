import {observable} from "mobx";

export interface ListInterface {
    name: string;
    index: number;
    isInEditMode: boolean;
    toggleEditMode: () => void;
    toJSON: () => ListJSON;
}

export interface ListJSON {
    name: string;
    index: number;
}

export class List implements ListInterface {
    @observable private _name: string;
    @observable private _inEditMode: boolean;
    private readonly _index: number;

    private static getRandomIndex() {
        return Math.floor((Math.random() * 10000) + 1);
    }

    constructor(name: string, index: number = List.getRandomIndex(), inEditMode: boolean = false) {
        this._name = name;
        this._inEditMode = inEditMode;
        this._index = index;
    }

    public set name(name: string) {
        this._name = name;
    }

    public toggleEditMode(): void {
        this._inEditMode = !this._inEditMode;
    }

    public get index(): number {
        return this._index;
    }

    public get name(): string {
        return this._name;
    }

    public get isInEditMode(): boolean {
        return this._inEditMode
    }

    public toJSON(): ListJSON {
        return {
            name: this.name,
            index: this.index
        }
    }

}