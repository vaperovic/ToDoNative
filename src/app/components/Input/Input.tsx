import React from 'react';
import { Component } from 'react';
import {
    Text,
    TextInput,
    TouchableHighlight,
    View,
    ViewStyle,
} from 'react-native';
import { InputStyle } from './Input.style';

export type InputProps = {
  addLabel: string;
  handleSave: (text: string) => void;
} & Partial<DefaultProps>;

interface InputState {
  text: string;
}

type DefaultProps = Readonly<typeof defaultProps>;

const defaultProps = {
  placeholder: '' as string,
  value: '' as string,
  style: {} as ViewStyle,
};

const EMPTY_STRING = '';

export class Input extends Component<InputProps, InputState> {

  public static defaultProps = defaultProps;

  constructor(props: InputProps) {
      super(props);
      this.state = {
          text: props.value || '',
        };
    }

  public createNewItem = async () => {
      const { handleSave } = this.props;
      const { text } = this.state;

      await handleSave(text);

      this.resetInput();
    }

  public handleOnChange = (text: string) => {
      this.setState({ text });
    }

  public render() {
      const { addLabel, placeholder, style } = this.props;
      const { text } = this.state;

      return (
            <View style={[InputStyle.inputContainer, style]}>
                <TextInput
                    value={text}
                    style={InputStyle.inputTextInput}
                    onChangeText={ this.handleOnChange }
                    onSubmitEditing={this.createNewItem}
                    placeholder={placeholder}
                />
                <TouchableHighlight
                    style={InputStyle.addButton}
                    onPress={this.createNewItem}>
                    <Text>{addLabel}</Text>
                </TouchableHighlight>
            </View>
        );
    }

  private resetInput = () => {
      this.setState({ text: EMPTY_STRING });
    }
}
