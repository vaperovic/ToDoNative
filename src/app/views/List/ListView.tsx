import React from 'react';
import { Component } from 'react';
import { View } from 'react-native';
import { inject, observer } from 'mobx-react';
import { NavigationStackScreenOptions } from "react-navigation";
import { Input } from "../../components/Input/Input";
import { ListsList } from "./components/ListsList/ListsList";
import {RootStoreInterface} from "../../../stores/Root.store";
import GlobalStyles from '../../styles';

export interface ListViewProps {
    navigation: any;
    rootStore: RootStoreInterface;
}

interface ListViewState {}

@inject('rootStore')
@observer
export class ListView extends Component<ListViewProps, ListViewState> {

    public static navigationOptions: NavigationStackScreenOptions = {
        title: 'My Lists',
    };

    private createList = async (name: string) => {
        const {rootStore} = this.props;

        await name && rootStore.listStore.createList(name);
    };

    public render() {
        const { navigation, rootStore } = this.props;

        return (
            <View style={GlobalStyles.fullHeightScreen}>
                <ListsList lists={rootStore.listStore.lists}
                           handleDelete={rootStore.listStore.deleteList}
                           handleEdit={rootStore.listStore.editList}
                           makeEditable={rootStore.listStore.toggleListEdit}
                           openDetails={navigation.navigate}
                />
                <Input placeholder={'New list item'}
                       addLabel={'Add list'}
                       handleSave={this.createList}
                />
            </View>
        );
    }
}