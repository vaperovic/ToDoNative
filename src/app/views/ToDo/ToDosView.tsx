import React from 'react';
import { Component } from 'react';
import { View } from 'react-native';
import { inject, observer } from 'mobx-react';

import { Input } from "../../components/Input/Input";
import { ToDosList } from "./components/ToDosList/ToDosList";
import { RootStoreInterface } from "../../../stores/Root.store";

import GlobalStyles from '../../styles';

export interface ToDosViewProps {
    rootStore: RootStoreInterface;
    navigation: any;
}

@inject('rootStore')
@observer
export class ToDosView extends Component<ToDosViewProps> {

    public static navigationOptions = ({ navigation }: any) => {
        return {
            title: navigation.getParam('parentList', undefined).name
        }
    };

    private addToDo = async (task: string) => {
        const { navigation, rootStore } = this.props;
        const parentList = navigation.getParam('parentList', undefined);

        task && await rootStore.toDoStore.createToDo(task, parentList.index);
    };

    public render() {
        const { navigation, rootStore } = this.props;
        const parentList = navigation.getParam('parentList', undefined);

        const todos = rootStore.toDoStore.getToDosForList(parentList.index);

        return (
            <View style={GlobalStyles.fullHeightScreen}>
                <ToDosList todos={todos}
                           handleDelete={rootStore.toDoStore.deleteToDo}
                           handleEdit={rootStore.toDoStore.editToDos}
                           makeEditable={rootStore.toDoStore.toggleToDoEdit}
                           toggleTaskFinish={rootStore.toDoStore.toggleToDoCompleted}
                />
                <Input placeholder={'New todo task'}
                       addLabel={'Add todo'}
                       handleSave={this.addToDo}
                />
            </View>
        );
    }
}