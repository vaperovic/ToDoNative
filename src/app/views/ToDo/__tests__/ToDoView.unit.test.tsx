import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import renderer from 'react-test-renderer';

import { ToDo } from '../../../../models/ToDo/ToDo.model';
import { ListStoreInterface } from '../../../../stores/List/List.store';
import { RootStoreInterface } from '../../../../stores/Root.store';
import { ToDoStoreInterface } from '../../../../stores/ToDo/ToDo.store';
import { Input } from '../../../components/Input/Input';
import { ToDosView } from '../ToDosView';

configure({ adapter: new Adapter() });

const ALL_TODOS = [{ index: 1, completed: false, name: 'todo', listId: 1 }];

const MockListStore = jest.fn<ListStoreInterface>(() => {});
const MockToDoStore = jest.fn<ToDoStoreInterface>(() => ({
  todos: [],
  createToDo: jest.fn((tekst: string, listId: number) => {}),
  deleteToDo: jest.fn(),
  toggleToDoEdit: jest.fn(),
  editToDos: jest.fn(),
  getToDosForList: jest.fn(() => {
      return ALL_TODOS.map(todo => (new ToDo(todo.name, todo.listId, todo.completed, todo.index)));
    }),
}));

const listStoreMock = new MockListStore();
const todosStoreMock = new MockToDoStore();

const MockRootStore = jest.fn<RootStoreInterface>(() => ({
  listStore: listStoreMock,
  toDoStore: todosStoreMock,
}));

const navigation = {
  getParam: jest.fn(() => {
      return {
          index: 1,
        };
    }),
};

const rootStoreMock = new MockRootStore();

describe('ToDoView unit test', () => {

  it('should render correctly', () => {

      const tree = renderer.create(<ToDosView rootStore={rootStoreMock} navigation={navigation}/>).toJSON();

      expect(tree).toMatchSnapshot();
    });

  it('should call createToDo when add todo clicked', () => {
      const wrapper = shallow(<ToDosView rootStore={rootStoreMock} navigation={navigation}/>).dive();

      wrapper.find(Input).first().props().handleSave('novi');

      expect(todosStoreMock.createToDo).toBeCalledTimes(1);
      expect(todosStoreMock.createToDo).toBeCalledWith('novi', 1);
      expect(navigation.getParam).toBeCalledTimes(3);
    });

});
