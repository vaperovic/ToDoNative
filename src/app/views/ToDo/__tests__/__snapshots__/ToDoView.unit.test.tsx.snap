// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoView unit test should render correctly 1`] = `
<View
  style={
    Object {
      "height": "100%",
    }
  }
>
  <RCTScrollView>
    <View>
      <RCTScrollView
        data={
          Array [
            Object {
              "completed": false,
              "index": 1,
              "listId": 1,
              "name": "todo",
            },
          ]
        }
        disableVirtualization={false}
        getItem={[Function]}
        getItemCount={[Function]}
        horizontal={false}
        initialNumToRender={10}
        keyExtractor={[Function]}
        maxToRenderPerBatch={10}
        numColumns={1}
        onContentSizeChange={[Function]}
        onEndReachedThreshold={2}
        onLayout={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        updateCellsBatchingPeriod={50}
        viewabilityConfigCallbackPairs={Array []}
        windowSize={21}
      >
        <View>
          <View
            onLayout={[Function]}
            style={null}
          >
            <View
              style={
                Object {
                  "borderBottomColor": "#eee",
                  "borderBottomWidth": 1,
                  "display": "flex",
                  "flexDirection": "column",
                  "paddingBottom": 10,
                  "paddingLeft": 10,
                  "paddingRight": 10,
                  "paddingTop": 10,
                }
              }
            >
              <Text
                style={
                  Array [
                    Object {
                      "textDecorationLine": "none",
                    },
                    Object {
                      "color": "#a7a7a7",
                      "fontSize": 18,
                    },
                  ]
                }
              >
                todo
              </Text>
              <View
                style={
                  Object {
                    "display": "flex",
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                    "width": "100%",
                  }
                }
              >
                <View>
                  <View
                    accessible={true}
                    isTVSelectable={true}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "alignItems": "center",
                        "backgroundColor": "transparent",
                        "justifyContent": "center",
                      }
                    }
                  >
                    <Text
                      allowFontScaling={false}
                      style={
                        Array [
                          Object {
                            "color": "black",
                            "fontSize": 24,
                          },
                          Object {
                            "backgroundColor": "transparent",
                          },
                          Object {
                            "fontFamily": "Material Icons",
                            "fontStyle": "normal",
                            "fontWeight": "normal",
                          },
                          Object {},
                        ]
                      }
                      testID="iconIcon"
                    >
                      
                    </Text>
                  </View>
                </View>
                <View>
                  <View
                    accessible={true}
                    isTVSelectable={true}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "alignItems": "center",
                        "backgroundColor": "transparent",
                        "justifyContent": "center",
                      }
                    }
                  >
                    <Text
                      allowFontScaling={false}
                      style={
                        Array [
                          Object {
                            "color": "black",
                            "fontSize": 24,
                          },
                          Object {
                            "backgroundColor": "transparent",
                          },
                          Object {
                            "fontFamily": "Material Icons",
                            "fontStyle": "normal",
                            "fontWeight": "normal",
                          },
                          Object {},
                        ]
                      }
                      testID="iconIcon"
                    >
                      
                    </Text>
                  </View>
                </View>
                <View
                  accessible={true}
                  isTVSelectable={true}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    Object {
                      "backgroundColor": "transparent",
                      "borderColor": "#ededed",
                      "borderRadius": 3,
                      "borderWidth": 0,
                      "margin": 5,
                      "marginBottom": 0,
                      "marginLeft": 10,
                      "marginRight": 10,
                      "marginTop": 0,
                      "opacity": 1,
                      "padding": 10,
                      "paddingBottom": 0,
                      "paddingTop": 0,
                    }
                  }
                >
                  <View
                    style={
                      Object {
                        "alignItems": "center",
                        "flexDirection": "row",
                      }
                    }
                  >
                    <Text
                      allowFontScaling={false}
                      style={
                        Array [
                          Object {
                            "color": "#bfbfbf",
                            "fontSize": 24,
                          },
                          Object {
                            "minWidth": 24,
                          },
                          Object {
                            "fontFamily": "FontAwesome",
                            "fontStyle": "normal",
                            "fontWeight": "normal",
                          },
                          Object {},
                        ]
                      }
                    >
                      
                    </Text>
                    <Text
                      style={
                        Object {
                          "color": "#43484d",
                          "fontWeight": "bold",
                          "marginLeft": 10,
                          "marginRight": 10,
                        }
                      }
                      testID="checkboxTitle"
                      theme={
                        Object {
                          "colors": Object {
                            "divider": "#bcbbc1",
                            "error": "#ff190c",
                            "grey0": "#393e42",
                            "grey1": "#43484d",
                            "grey2": "#5e6977",
                            "grey3": "#86939e",
                            "grey4": "#bdc6cf",
                            "grey5": "#e1e8ee",
                            "greyOutline": "#bbb",
                            "primary": "#2089dc",
                            "searchBg": "#303337",
                            "secondary": "#8F0CE8",
                          },
                        }
                      }
                    >
                      Finish
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </RCTScrollView>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderTopColor": "#eee",
          "borderTopWidth": 1,
          "display": "flex",
          "flexDirection": "row",
        },
        Object {},
      ]
    }
  >
    <TextInput
      allowFontScaling={true}
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      placeholder="New todo task"
      style={
        Object {
          "flex": 8,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "borderLeftColor": "#eee",
          "borderLeftWidth": 1,
          "display": "flex",
          "flex": 2,
          "flexDirection": "row",
          "height": "70%",
          "paddingLeft": 10,
        }
      }
    >
      <Text
        style={null}
      >
        Add todo
      </Text>
    </View>
  </View>
</View>
`;
